### PR TITLE
Add CDI bean to allow easy access to some hibernate metadata at runtime

### LINKED ISSUE #9049: NPE when injecting EntityManager/EntityManagerFactory into ObjectMapperCustomizer
**Describe the bug**
I implemented the Jackson [ObjectMapperCustomizer](https://quarkus.io/guides/rest-json#configuring-json-support) and tried to @Inject the Hibernate EntityManager or EntityManagerFactory.
On quarkus:dev startup, a NullPointerException gets thrown:

```
Caused by: java.lang.NullPointerException
	at io.quarkus.agroal.runtime.AbstractDataSourceProducer.getDataSourceBuildTimeConfig(AbstractDataSourceProducer.java:375)
```

**Expected behavior**
Application should start, EM/EMF should get injected.

**Actual behavior**
NPE, see description

**To Reproduce**
1. Check out https://github.com/HonoluluHenk/quarkus-bug-jacksoncustomizer-hibernate.git
2. ./mvnw clean test
3. Watch it crash&burn

**Configuration**
N/A

**Screenshots**
N/A

**Environment (please complete the following information):**
 - Quarkus version or git rev: 1.4.1.Final
 

**Additional context**
Most specific stacktrace entry:
```
Caused by: java.lang.NullPointerException
        at io.quarkus.agroal.runtime.AbstractDataSourceProducer.getDataSourceBuildTimeConfig(AbstractDataSourceProducer.java:375)
        at io.quarkus.agroal.runtime.DataSourceProducer.createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595(DataSourceProducer.zig:31)
        at io.quarkus.agroal.runtime.DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.create(DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.zig:354)
        at io.quarkus.agroal.runtime.DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.create(DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.zig:297)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:79)
        at io.quarkus.arc.impl.ComputingCache$CacheFunction.lambda$apply$0(ComputingCache.java:99)
        at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
        at io.quarkus.arc.impl.ComputingCache.getValue(ComputingCache.java:41)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:25)
        at io.quarkus.agroal.runtime.DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.get(DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.zig:269)
        at io.quarkus.agroal.runtime.DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.get(DataSourceProducer_ProducerMethod_createDataSource_4c279770c59fa93dcf179ab15a1363f01f14f595_5923a8feaf2f425c7e2697f04d550dce98568e14_Bean.zig:41)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:380)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:393)
        at io.quarkus.arc.impl.ArcContainerImpl.instanceHandle(ArcContainerImpl.java:363)
        at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:200)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.injectDataSource(FastBootHibernatePersistenceProvider.java:265)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(FastBootHibernatePersistenceProvider.java:155)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:48)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:110)
        at io.quarkus.hibernate.orm.runtime.JPAConfig.getEntityManagerFactory(JPAConfig.java:43)
        at io.quarkus.hibernate.orm.runtime.JPAResourceReferenceProvider.lambda$get$0(JPAResourceReferenceProvider.java:26)
        at io.quarkus.arc.impl.ResourceProvider.get(ResourceProvider.java:36)
        at io.quarkus.hibernate.orm.runtime.DefaultEntityManagerFactoryProducer_Bean.create(DefaultEntityManagerFactoryProducer_Bean.zig:61)
        ... 78 more

```

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #9049

### CODE REVIEW

**Comment by gsmet** [comment]
I wonder if it's a bit too early to do that. Or at least, we should try to consider having multiple persistence units in the design.

**Comment by geoand** [comment]
Are persistence units named? I am thinking that something similar to Datasource configuration could work here perhaps. The existing method would work for the default persistence unit and then the beans of named persistence units would be returned in another method. Would that work?

**Comment by Sanne** [comment]
I'm still skeptical on the reasons to need this.  I've seen mentiond "Jackson serialization customizations at runtime" .. that doesn't feel right, I had suggested on  #9049 to process such things at build time.

Could we at least make sure that such things have zero overhead when they are not used?

**Comment by geoand** [comment]
> Could we at least make sure that such things have zero overhead when they are not used?

That's exactly what `HibernateMetadataWithoutInjectionTest` does. When a class isn't used in an injection, Arc doesn't create a bean for it (cc @mkouba)

**Comment by Sanne** [comment]
Also.. IFF there are entities being managed, then you'll be able to access such metadata from the EntityManager. I have not understood why one couldn't just check if the EntityManager is indeed available (if there isn't one then there's not going to be any managed entities either).

To answer your question: yes persistence units have a name.

**Comment by geoand** [comment]
> I'm still skeptical on the reasons to need this. I've seen mentiond "Jackson serialization customizations at runtime" .. that doesn't feel right, I had suggested on #9049 to process such things at build time.
> 

Problem is that users might need to do arbitrary things with this list. Not mention that for each extension that could possible interract with this data, we would need to create custom integration points with user code. 
IMHO, CDI is that injection point and we should try and leverage it as much as possible

**Comment by geoand** [comment]
> Also.. IFF there are entities being managed, then you'll be able to access such metadata from the EntityManager. I have not understood why one couldn't just check if the EntityManager is indeed available (if there isn't one then there's not going to be any managed entities either).
> 

Because you might need to be able to do something with CDI before the persistence unit is ready. That's essentially what #9049 was all about.

I am not dead set on this by any means. If you think that it doesn't make sense to have it or that it cause us problems down the road, I have no problem droping it.

> To answer your question: yes persistence units have a name.

So you agree with @gsmet that if we provide the feature that it should take multiple persistence units into account now, even though we don't yet have that support in Quarkus?

**Comment by Sanne** [comment]
> Problem is that users might need to do arbitrary things with this list. Not mention that for each extension that could possible interract with this data, we would need to create custom integration points with user code.

Any extension can either dig the Jandex index itself, or consume the build Items that the Hibernate ORM will produce - it actualy produces them for such purposes.

But more importantly, isn't _anything_ one might want to do with such a list a concern related to metadata and therefore related to introspection and build concerns?

I guess I'm just concerned that such capabilities don't belong into the runtime JVM as it's not aligned with our general strategy.

**Comment by Sanne** [comment]
> So you agree with @gsmet that if we provide the feature that it should take multiple persistence units into account now, even though we don't yet have that support in Quarkus?

If you could, that would be great :)

  1. We need to eventually be able to support multi-PU without breaking such new APIs
  2. It's inconvenient that introducing other minor features makes the work for multi-PU much larger and harder

**Comment by geoand** [comment]
For case in hand, a user wants to configure Jackson in some specific way for each entity. Sure the user could write an extension to do that, but is really the right abstraction?
Jackson gets configured at runtime anyway, so the best we could do is generate some scaffolding class that the user would override. That doesn't seem to provide much value however.

**Comment by Sanne** [comment]
Well if there's no overhead at runtime I won't block the PR :)

but it does feel like something that needs its own extension. Not least, serializing entities via Jackson will cause many more issues so I suspect this would need an extension.

I wasn't aware that Jackson would be "configured at runtime" - I don't know much about it but it just makes me think there's another gap there? We should fix the gaps rather than adding more.

**Comment by geoand** [comment]
> Well if there's no overhead at runtime I won't block the PR :)
> 
> but it does feel like something that needs its own extension. Not least, serializing entities via Jackson will cause many more issues so I suspect this would need an extension.
> 
> I wasn't aware that Jackson would be "configured at runtime" - I don't know much about it but it just makes me think there's another gap there? We should fix the gaps rather than adding more.

That's an interesting idea. We could certainly try and follow the same strategy as with Hibernate and provide some slimmed down version of it, but given the various usages we've seen people adopt with it, I kind of doubt we could provide the functionality users come to expect from it.

**Comment by Sanne** [comment]
Ok yes I understand that I'm not being very pragmatic and I don't want to get in the way.

Let's separate my idealistic goals from the needs of today - but I'm only voicing such ideas so to start a brainstorming for future improvements.

Consider me +1 on this provided that:
 - you keep the multi-persistence unit requirement in mind (if not solve it, at least make sure it will be solveable)
 - document in some way that this API is experimental and is unlikely to stay around for long

Especially as Jackson "configured at runtime" will needs reflection and possibly other itches, there's many things we could do better bu having better build-time support.

**Comment by geoand** [comment]
@Sanne Thanks!

My latest push should make it trivial to update this new API when multiple PUs are supported. The names of the methods could perhaps be better but I am notoriously bad at naming...

**Comment by mkouba** [comment]
>  When a class isn't used in an injection, Arc doesn't create a bean for it...

Well, the bean is not included but if your deployment contains a lot of entities then the recorded bytecode is still there and the result could be a very looong method ;-). Isn't there a way to query some Hibernate internals to get this info? Something like https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/ArcEndpointRecorder.java#L31

**Comment by geoand** [comment]
> > When a class isn't used in an injection, Arc doesn't create a bean for it...
> 
> Well, the bean is not included but if your deployment contains a lot of entities then the recorded bytecode is still there and the result could be a very looong method ;-). Isn't there a way to query some Hibernate internals to get this info? Something like https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/ArcEndpointRecorder.java#L31

There is, but requires that the `EntityManager` is already setup

File: extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/HibernateMetadata.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Returns a map where the persistence unit name is the key and the class names of the associated entities are the value
+     */
+    Map<String, Set<String>> entitiesByPersistenceUnit();
+
+    /**
+     * Returns a set of class names of the entities that are associated with the default persistence unit
+     */
+    default Set<String> entitiesOfDefaultPersistenceUnit() {
+        return entitiesByPersistenceUnit().getOrDefault(DEFAULT_PERSISTENCE_UNIT_NAME, Collections.emptySet());
```
Comment (gsmet) [inline extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/HibernateMetadata.java:34]: I was wondering if a `PersistenceUnit` container would be useful instead?

You would have `getPersistenceUnit(String)` and `getDefaultPersistenceUnit()` which would return a `PersistenceUnit` and you would have a `getEntities()` in this container.

Comment (geoand) [reply extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/HibernateMetadata.java:34]: Interesting, I like it.

I assume you that the new class would be called `PersistenceUnitMetadata` or something?

Comment (geoand) [reply extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/HibernateMetadata.java]: PR updated to reflect this design

**Comment by gsmet** [comment]
@geoand this is outdated. But the good news is that we now have real multiple persistence units support :).

Do you want to revisit this one? Or should we close it given it didn't get a lot of traction?

**Comment by geoand** [comment]
I'd say let's close it for now. We can always revisit if needed

**Pull request closed without merging**